MapsetVerifier.Checks/Taiko/Compose/UnrankableFinisherCheck.cs

```diff
@@ -0,0 +1,195 @@
+            };
+
+            // any finisher pattern spacing equal to or smaller than this gap is a problem
+            var maximalGapBeatsWarning = new Dictionary<Beatmap.Difficulty, double>()
+            {
+                { Beatmap.Difficulty.Hard,    1.0/3 }
+            };
+
+            // any finisher pattern spacing equal to or smaller than this gap without a color change before is a problem
+            var maximalGapBeatsRequiringColorChangeBefore = new Dictionary<Beatmap.Difficulty, double>()
+            {
+                { Beatmap.Difficulty.Insane,     1.0/4 }
+            };
+
+            // any finisher pattern spacing equal to or smaller than this gap without a color change after is a problem
+            var maximalGapBeatsRequiringColorChangeAfter = new Dictionary<Beatmap.Difficulty, double>() { };
+
+            // any finisher pattern spacing equal to or smaller than this gap without a color change before is a warning
+            var maximalGapBeatsRequiringColorChangeBeforeWarning = new Dictionary<Beatmap.Difficulty, double>()
+            {
+                { Beatmap.Difficulty.Expert,   1.0/3 },
+                { Beatmap.Difficulty.Ultra,    1.0/3 }
+            };
+
+            // any finisher pattern spacing equal to or smaller than this gap without a color change after is a warning
+            var maximalGapBeatsRequiringColorChangeAfterWarning = new Dictionary<Beatmap.Difficulty, double>()
+            {
+                { Beatmap.Difficulty.Expert,   1.0/3 },
+                { Beatmap.Difficulty.Ultra,    1.0/3 }
+            };
+
+            // any finisher pattern spacing equal to or smaller than this gap while not being at the end of the pattern is a problem
+            var maximalGapBeatsRequiringFinalNote = new Dictionary<Beatmap.Difficulty, double>()
+            {
+                { Beatmap.Difficulty.Insane,     1.0/4 }
+            };
+
+            // any finisher pattern spacing equal to or smaller than this gap while not being at the end of the pattern is a warning
+            var maximalGapBeatsRequiringFinalNoteWarning = new Dictionary<Beatmap.Difficulty, double>()
+            {
+                { Beatmap.Difficulty.Expert,   1.0/3 },
+                { Beatmap.Difficulty.Ultra,    1.0/3 }
+            };
+
+            foreach (var diff in difficulties)
+            {
+                foreach (var current in finishers)
+                {
+                    var sameColorBefore = current.IsMono();
+                    var sameColorAfter = current.Next()?.IsMono() ?? false;
+                    var isInPattern = !current.IsNotInPattern();
+                    var isFirstNote = current.IsAtBeginningOfPattern();
+                    var isFinalNote = current.IsAtEndOfPattern();
+
+                    // check for unrankable finishers (problem)
+                    if ((CheckGap(diff, maximalGapBeats, beatmap, current) && isInPattern) ||
```

Comment:
While this might look cleaner, having the `CheckGap` at the end of each sequence of checks is actually better to avoid unneeded calculations, given that you already have all the variables determined for `sameColorBefore`, `sameColorAfter`, `isInPattern`, etc.

Reply:
That's true. I think I'm personally more of a fan of how the current reads though - seems easier to follow along, despite the performance cost. I guess will leave up to hivie